Async bo rule and notebook

apaleyes opened the pull request:
This PR replaces and expands #363.

Here we:
* Make changes in greedy acquisition function related to batch initialization and handling of pending points
* Introduce new AsyncEGO rule.
* Provide a notebook that illustrates how one could do async BO with Trieste

Unit tests, docs and other trimmings to come at a later point, I am mostly looking for feedback at this point. One note about new rule is that, unlike TrustRegion, this one works not with other rules but with function builders directly. Happy to hear suggestions of how to change that.

**Update:** all the trimmings are now done. The notebook is decorated with many-many comments to make it as accessible as possible, very happy to improve it further if people still find it unclear what's all this multiprocessing stuff is about.

apaleyes committed:
First demo of async bo

apaleyes committed:
Passing through pending points

apaleyes committed:
new output

apaleyes committed:
Add plotting to notebook

apaleyes committed:
Update notebook

apaleyes committed:
Use pending points in prepare

apaleyes committed:
Added some comments

apaleyes committed:
Move processing of pending points to a separate function, call it in prepare

apaleyes committed:
Formatting, tests

apaleyes committed:
rollback some early changes

apaleyes committed:
format black

apaleyes committed:
Merge remote-tracking branch 'upstream/develop' into pending-points

apaleyes committed:
Revert some changes to prepare, update notebook

apaleyes committed:
Removed ipynb

apaleyes committed:
Remove optional

apaleyes committed:
Docstring for AsyncEGO

uri-granta commented on trieste/acquisition/rule.py:
```diff
@@ -200,14 +202,144 @@ def acquire(
 
 
+class AsyncEGO(AcquisitionRule[State[Optional["AsyncEGO.State"], TensorType], SP_contra]):
+    """AsyncEGO rule, as name suggests, is designed for async BO scenarios.
```
```suggestion
    """AsyncEGO rule, as name suggests, is designed for asynchronous BO scenarios.
```

henrymoss replied:
Give it a proper name! AsynchronousEfficientGlobalOptimization

apaleyes replied:
It had full name initially, but I've cut it down because it just felt too long and hard to read. Do you think I shall unfold it back?

apaleyes replied:
eventually settled on AsyncEfficientGlobalOptimization

uri-granta commented on trieste/acquisition/rule.py:
```diff
@@ -200,14 +202,144 @@ def acquire(
+    we want to immediately use acquisition function to launch a new observation and avoid wasting computational resources.
+
+    To make the best decision about next point to observe, acquisition function needs to be aware of currently running observations.
+    We call such points "pending". AsyncEGO provides a ``AsyncEGO.State`` object that keeps track of pending points
```
I think we should differentiate between "pending" points used by greedy acquisition functions (points that have been selected but haven't been executed yet) and "outstanding" points, which have been executed but haven't returned.

```suggestion
    We call such points "outstanding". AsyncEGO provides a ``AsyncEGO.State`` object that keeps track of outstanding points
```

apaleyes replied:
Can you elaborate a bit what's the difference between pending and outstanding? I don't think I understand.

Once the point was chosen by the acquisition function, there is no difference any more, as I understand it. It could have been chosen 5 calls ago or just now, it is still treated exactly the same - as a point which was selected for observations, but result for which hasn't arrived yet.

uri-granta replied:
It's like the difference between things in your Amazon basket, and things that you've ordered but haven't arrived yet.

* Greedy acquistiion functions choose points one at a time. The first points are pending until we've selected the entire batch.
* Asyn BO choose some points (greedily) and then ask for them to be evaluated (like normal BO), but then starts another loop before all of them have finished executing. The ones that haven't finished yet are outstanding.

uri-granta commented on trieste/acquisition/rule.py:
```diff
@@ -200,14 +202,144 @@ def acquire(
+    ):
+        """
+        :param builder: Acquisition function builder. Only greedy batch approaches are supported,
+            because they are the only ones currently supporting the pending points concept.
```
```suggestion
            because they can be told what points are outstanding.
```

apaleyes replied:
fixed, except for outstanding/pending

uri-granta commented on trieste/acquisition/rule.py:
```diff
@@ -200,14 +202,144 @@ def acquire(
+            )
+
+        if builder is None:
+            raise ValueError("Please specify an acquisition builder")
```
Might be nice to default to `LocalPenalizationAcquisitionFunction(search_space)` but this requires knowing the global search space. @henrymoss: opinion?

henrymoss replied:
Its nice to have a default for sure, but probably too difficult in this case (without passing search space). Once we have a greedy Thompson sampling method, perhaps that could be the default?

apaleyes replied:
no default for now

uri-granta commented on trieste/acquisition/rule.py:
```diff
@@ -200,14 +202,144 @@ def acquire(
+                mask = tf.map_fn(is_not_in_dataset, pending_points, fn_output_signature=tf.bool)
+                pending_points = tf.boolean_mask(pending_points, mask)
+
+            acquisition_function = self._builder.prepare_acquisition_function(
```
Like EGO, we should store the acquisition function in `self._acquisition_function` and update it if it already exists to avoid retracing unnecessarily.

apaleyes replied:
I am a bit vary of closures here updating the `self._acquisition_function` inside the state callable. But will play around and see if that works

apaleyes replied:
tried it, and it seems to work fine

uri-granta commented on trieste/acquisition/rule.py:
```diff
@@ -200,14 +202,144 @@ def acquire(
+        To keep them up to date, pending points are compared against the given dataset,
+        and whatever points are in the dataset are deleted.
+        Then the current batch is generated by calling the acquisition function,
+        and all points in the batch are added to the known pending points.
```
Nice clear docstring, thanks!

apaleyes replied:
cheers!

uri-granta commented on trieste/acquisition/rule.py:
```diff
@@ -200,14 +202,144 @@ def acquire(
+            points from the previous acquisition state.
+        """
+
+        def state_func(state: AsyncEGO.State | None) -> tuple[AsyncEGO.State | None, TensorType]:
```
We normally call the state functions `go` for some reason (e.g. see TrustRegion).

```suggestion
        def go(state: AsyncEGO.State | None) -> tuple[AsyncEGO.State | None, TensorType]:
```

apaleyes replied:
Since there is only one such instance so far, I dare challenging it! `go` is rather cryptic, I find `state_func` clearer. If anything, probably we shall update TrustRegion?

henrymoss replied:
If we make this change, lets make it consistent across states

apaleyes replied:
i made the change in TrustRegion too

uri-granta commented on trieste/acquisition/rule.py:
```diff
@@ -200,14 +202,144 @@ def acquire(
+
+        def state_func(state: AsyncEGO.State | None) -> tuple[AsyncEGO.State | None, TensorType]:
+            @tf.function
+            def is_not_in_dataset(x):
```
```suggestion
            def is_not_in_dataset(x : TensorType) -> TensorType:
```

apaleyes replied:
fixed

uri-granta commented on trieste/acquisition/rule.py:
```diff
@@ -200,14 +202,144 @@ def acquire(
+    We call such points "pending". AsyncEGO provides a ``AsyncEGO.State`` object that keeps track of pending points
+    """
+    @dataclass(frozen=True)
+    class State:
```
What's the reasoning behind using a frozen dataclass for the state, rather than just the pending_points/outstanding_points?

apaleyes replied:
No good reason. I anticipated more stuff in the state, but ended up with only one field.

apaleyes replied:
Actually thinking about it a bit more we shall probably stick to a class for a state, even if it has just one property. It's a bit more extensible and looks more reasonable in types. Or a namedtuple.

uri-granta commented on trieste/acquisition/rule.py:
```diff
@@ -200,14 +202,144 @@ def acquire(
+            points = self._optimizer(search_space, acquisition_function)
+            pending_points = points
+
+            for _ in range(self._num_query_points - 1):  # greedily allocate batch elements
```
```suggestion
            for _ in range(self._num_query_points - 1):  # greedily allocate additional batch elements
```

apaleyes replied:
fixed

uri-granta commented on trieste/acquisition/function.py:
```diff
@@ -1289,6 +1290,9 @@ def update_acquisition_function(
         :param models: The models over each dataset in ``datasets``.
         :param pending_points: Points already chosen to be in the current batch (of shape [M,D]),
             where M is the number of pending points and D is the search space dimension.
+        :param new_optimization_step: Indicates if this call to update_acquisition_function
```
```suggestion
        :param new_optimization_step: Indicates whether this call to update_acquisition_function
```

henrymoss replied:
I think this is a good change. Certainly better than the implicit behaviour we had before

apaleyes replied:
"explicit is better than implicit" The Zen of Python

uri-granta commented on trieste/acquisition/function.py:
```diff
@@ -1289,6 +1290,9 @@ def update_acquisition_function(
         :param pending_points: Points already chosen to be in the current batch (of shape [M,D]),
             where M is the number of pending points and D is the search space dimension.
+        :param new_optimization_step: Indicates if this call to update_acquisition_function
+            is to start of a new optimization step, of to continue collecting batch of points
```
```suggestion
            is to start of a new optimization step, or to continue collecting batch of points
```

apaleyes replied:
fixed

hstojic replied:
I don't see its fixed?

uri-granta commented on trieste/acquisition/function.py:
```diff
@@ -1443,37 +1461,54 @@ def prepare_acquisition_function(
-        # no penalization required if no pending_points.
-        return self._update_base_acquisition_function(dataset, model)
+        acq = self._update_base_acquisition_function(dataset, model)
+        if pending_points is not None:
```
Not sure whether this change is actually necessary, as the very first time pending_points should always be None right? Still, if we do make the change then we should probably also handle passing in empty pending_points.

```suggestion
        if pending_points is not None or len(pending_points) == 0:
```

apaleyes replied:
I guess that's an attempt at being general. Consider some complex ask/tell scenarios where optimizer instance is created anew for each ask. I agree that's rather hypothetical, but it seems easy to support, hence this change. Will cover the empty tensor case, thanks.

uri-granta reviewed (commented):
Overall approach looks good! Various comments. Also you'll need to add unit and integration tests for AsyncEGO (and maybe a couple of unit tests for the function changes).

henrymoss commented on docs/notebooks/async_bo_demo.py:
```diff
@@ -0,0 +1,326 @@
+sync_bo = AskTellOptimizer(search_space, initial_data, model, local_penalization_rule)
+
+
+# thread-safe queues
+m = Manager()
+points_q = m.Queue()
+observations_q = m.Queue()
+observer_processes = []
+pid = os.getpid()
+
```
I dont really understand any of the the multi processing stuff. This will need to be better documented in the final version!

apaleyes replied:
did my best to explain what's going on, let me know if you feel it needs more of anything

henrymoss commented on docs/notebooks/async_bo_demo.py:
```diff
@@ -0,0 +1,326 @@
+    for i in range(num_iterations):
+        print(f"Process {pid}: Main     : iteration {i} starts", flush=True)
+
+        # get a batch of points, send them to queue
+        # each worker picks up a point and processes it
```
does this give a batch or a single point?

henrymoss commented on docs/notebooks/async_bo_demo.py:
```diff
@@ -0,0 +1,326 @@
+        observer_processes.append(observer_p)
+
+    # init the queue with first batch of points
+    for _ in range(num_workers):
+        point = bo.ask()
+        pq.put(np.atleast_2d(point.numpy()))
+
+
+    while points_observed < len(sync_lp_observations) - len(initial_data):
+        try:
+            new_data = oq.get_nowait()
+            print(f"Process {pid}: Main     : received data {new_data}", flush=True)
+        except:
+            continue
+        
+        new_data = Dataset(query_points=tf.constant(new_data[0], dtype=tf.float64),
+                           observations=tf.constant(new_data[1], dtype=tf.float64))
+
+        bo.tell(new_data)
+        point = bo.ask()
+        print(f"Process {pid}: Main     : acquired point {point}", flush=True)
+        pq.put(np.atleast_2d(point))
+        points_observed += 1
```
This looks like the right logic :)

henrymoss reviewed (commented):
LGTM

apaleyes committed:
Formatting, tests

apaleyes committed:
Many comments in the notebook, tests, new rule name

apaleyes committed:
remove actual ipynb notebook

hstojic commented on trieste/acquisition/rule.py:
```diff
@@ -200,14 +202,165 @@ def acquire(
+
+    def __repr__(self) -> str:
+        """"""
+        return f"""AsyncEGO(
```
perhaps AsyncEfficientGlobalOptimization rather than AsyncEGO?

uri-granta replied:
Yes, this should match the class name.

apaleyes replied:
of course, thanks

hstojic commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+# %% [markdown]
+# # Asynchronous Bayesian optimization with Trieste
+#
+# In this notebook we show how Bayesian optimization can be done asynchronuosly. That is pertinent in scenarios when the objective function we are aiming to optimize can be run for several points in parallel, and observations might return back at different times. To avoid wasting resources we immediatelly request the next point asynchronuosly, taking into account points that still being evaluated ...
```
that still being evaluated -> that are still being evaluated

hstojic replied:
if you want to add an example here: when optimising hyperparameters of ML models its realistic that for some configurations training time will be much longer

henrymoss replied:
In this notebook we demonstrate Trieste's ability to perform asynchronous Bayesian optimisation, as is suitable for scenarios where the objective function  can be run for several points in parallel but where observations might return back at different times. To avoid wasting resources waiting for the evaluation of the whole batch, we immediately request the next point asynchronously, taking into account points that still being evaluated.

hstojic commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+
+
+# %% [markdown]
+# First, let's define a simple objective that will emulate workload of variable time. We will be using class BO benchmark function [Branin](https://www.sfu.ca/~ssurjano/branin.html), and insert sleep call in the middle of the calculation to emulate delay. Our sleep delay is a scaled sum of all input values to make sure delays are uneven.
```
class BO -> classic Bayesian optimisation?

henrymoss replied:
First, let's define a simple objective that will emulate evaluations taking variable time. We will be using a classic BO benchmark function [Branin](https://www.sfu.ca/~ssurjano/branin.html) with a sleep call inserted in the middle of the calculation to emulate delay. Our sleep delay is a scaled sum of all input values to make sure delays are uneven.

apaleyes replied:
that was silly of me to miss

hstojic commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+    if points.shape[1] != 2:
+        raise ValueError(f"Incorrect input shape, expected (*, 2), got {points.shape}")
+
+    def scaled_branin(x):
```
it should be possible to wrap the existing function that we have in `objectives` here rather than repeating the code, no?

apaleyes replied:
good point, fixed

hstojic commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+    gpr = gpflow.models.GPR(data.astuple(), kernel, noise_variance=1e-5)
+    gpflow.set_trainable(gpr.likelihood, False)
+
+    return GPflowModelConfig(
```
perhaps we could simplify here and use `GaussianPreocessRegression` wrapper instead?

apaleyes replied:
good call

uri-granta commented on trieste/acquisition/rule.py:
```diff
@@ -45,6 +45,9 @@
 T_co = TypeVar("T_co", covariant=True)
 """ Unbound covariant type variable. """
 
+S = TypeVar("S")
```
I'm probably being blind, but where is this used?

apaleyes replied:
You are not, it isn't used. It's a remnant of my attempts of writing general StatefulEGO. Removed

hstojic commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+def observer_proc(points_queue, observations_queue):
+    pid = os.getpid()
+
+    while True:
```
I'm not sure I understand why is loop necessary here - this function is used for a single observation no? I see why it would be needed if you are running it until there are no more points, but then description above says single observation, so I'm not sure...

apaleyes replied:
It isn't. This worker sits in a separate process and listens to the queue. When a new point arrives, it processes it, puts the output in the queue, and sits waiting again. In the end we terminate it to clean up resources.

What you have in mind is a different way to implement it: firing workers for each new point and taking them down immediately after the observation is done. It's also a reasonable approach, I've just implemented a different one here.

hstojic commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+# %%
+# Number of worker processes to run simultaneously
+# Setting this to 1 will turn both setups into non-batch sequential optimization
+num_workers = 3
```
perhaps set to 2, enough to illustrate the point and will reduce computation time?

apaleyes replied:
Actually setting this to 2 will increase the computation time. More workers - faster compute.

uri-granta reviewed (approved):
Looks good (though my preference would still be to distinguish in name between the greedy pending points and async pending points)

hstojic commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+
+# setup Ask Tell BO
+model_spec = build_model(initial_data)
+model = create_model(model_spec)
```
unnecessary if you use the model wrapper directly as suggested above

apaleyes replied:
yes, removed

hstojic commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+# ## Asynchronous optimization
+# This section runs the asynchronous optimization routine. We first setup the [ask/tell optimizer](ask_tell_optimization.ipynb). Notice that we use **AsyncEfficientGlobalOptimization** rule specifically designed for asynchronous scenarios. Next we create thread-safe queues for points and observations, and run the optimization loop.
+#
+# Crucially, even though we are using batch acquisition function Local Penalization, we specify batch size of 1. This is because we don't really want a batch. Since the amout of workers we have is fixed, whenever we see a new observation we only need one point back. However this process can only be done with acquisition funcitons that implement greedy batch collection strategies, because they are ...
```
amout -> amount

hstojic replied:
why do we then even allow setting `num_query_points` to anything else than 1?

henrymoss replied:
I agree. There is no need to allow batches for asynchronous

apaleyes replied:
agreed and removed after some further thinking

hstojic commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+
+# %% [markdown]
+# ## Asynchronous optimization
+# This section runs the asynchronous optimization routine. We first setup the [ask/tell optimizer](ask_tell_optimization.ipynb). Notice that we use **AsyncEfficientGlobalOptimization** rule specifically designed for asynchronous scenarios. Next we create thread-safe queues for points and observations, and run the optimization loop.
```
additional sentence on why ask-tell is needed here would be useful

apaleyes replied:
added

hstojic commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+points_observed = 0
+
+start = timeit.default_timer()
+try:
```
a bit of education for me - can you clarify why is `try` needed here?

apaleyes replied:
to clean up running processes in case anything goes wrong and the code falls with exception. Note that we are not actually using `except`, only `try/finally`

hstojic commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+# To compare outcomes of sync and async runs, let's plot their respective regrets side by side, and print out the running time. We expect async scenario to run a little bit faster for this toy problem.
+
+# %%
+from trieste.objectives import SCALED_BRANIN_MINIMUM
```
imported already, right?

apaleyes replied:
yup! removed

hstojic commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+#
+# In this notebook we show how Bayesian optimization can be done asynchronuosly. That is pertinent in scenarios when the objective function we are aiming to optimize can be run for several points in parallel, and observations might return back at different times. To avoid wasting resources we immediatelly request the next point asynchronuosly, taking into account points that still being evaluated ...
+#
+# To contrast this approach with regular [batch optimization](batch_optimization.ipynb), this notebook also shows how to run parallel synchronous batch approach.
```
it would be good to add a sentence on potential performance differences, async one will have more information as some observations would have had arrived and model could be improved, right?

apaleyes replied:
I don't know if we can clearly make this call. Although [this paper](https://arxiv.org/abs/1901.10452) claims async is actually more efficient in both runtime and sample efficiency, i wouldn't rush to set user expectations in general. Any ideas how to phrase it without claiming much?

hstojic replied:
I think its enough to put a stress on "potential" and give usual "depending on the use case" etc - I do think its important to say smth about why would you use it - link to that paper would also be good to add

apaleyes replied:
Added something that is hopefully vague enough

hstojic commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+
+# %% [markdown]
+# ## Comparison
+# To compare outcomes of sync and async runs, let's plot their respective regrets side by side, and print out the running time. We expect async scenario to run a little bit faster for this toy problem.
```
a comment on expected differences in performance?

apaleyes replied:
see reply above

hstojic reviewed (commented):
what I could assess looks very good - mostly notebook, multiprocessing in python an AsyncEfficientGlobalOptimization itself I canno really evaluate

I have left lots of small suggestions for improvement in the notebook, nothing major

hstojic commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+from multiprocessing import Manager, Process
+
+# %% [markdown]
+# We now define several common functions to implement the described setup. First we define a worker function that will be running a single observation in a separate process. Worker takes both queues as an input, reads next point from the points queue, makes an observation, and inserts observed data into the observations queue.
```
not sure if its needed, but perhaps it would be useful to add something about assumptions - that the user actually has multiple threads/cores

perhaps add a link to a useful python documentation on multiprocessing

apaleyes replied:
No assumptions really, that code can run anywhere  Python can. But in case of 1 CPU there will be no performance gain, in fact it may degrade. I'll add a short note in the last section.

hstojic commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+# %% [markdown]
+# ## Multiprocessing setup
+#
+# To keep this notebook as reproducible as possible, we will only be using Python's multiprocessing package here. In this section we will explain our setup and define some common code to be used later.
```
perhaps mention in passing what would be alternatives to Python multiprocessing package?

apaleyes replied:
My plan is to provide a second notebook that will use one of the alternatives. Then we'll link the two.

henrymoss commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+#
+# To keep this notebook as reproducible as possible, we will only be using Python's multiprocessing package here. In this section we will explain our setup and define some common code to be used later.
+#
+# In both synchronous and asynchronous scenarios we will have a fixed set of worker processes performing observations. We will also have a main process responsible for optimization process with Trieste. When Trieste suggests a new point, it is inserted into a points queue. One of the workers picks this point from the queue, performs the observation, and inserts the output into observations queue. ...
```
This section is much clearer!

henrymoss replied:
In both synchronous and asynchronous scenarios we will have a fixed number of worker processes collecting observations. We will also have a main process responsible for thee Bayesian optimization process through Trieste. When Trieste suggests a new point, it is inserted into a points queue. One of the workers picks this point from the queue, performs the observation, and inserts the output into observations queue. The main process then picks up the observation from the queue, at which moment it either waits for the rest of the points in the batch to come back (synchronous scenario) or immediately suggests a new point (asynchronous scenario). This process continues either for a certain number of iterations or until we accumulate necessary number of observations.

henrymoss commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+num_workers = 3
+# Number of iterations to run the sycnhronous scenario for
+num_iterations = 10
+# Number of observations to collect in asynchronous scenario
```
in the asynchronous scenario

henrymoss commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+num_iterations = 10
+# Number of observations to collect in asynchronous scenario
+num_observations = num_workers * num_iterations
+# Set this flag to False to disable sleep delays in case you want the notebook to execute quickly
+enable_sleep_delays = True
```
Is there a way to set this to be False, when we run our notebook tests? i.e checking an environment variable? @hstojic @uri-granta ?

henrymoss replied:
We could do something similar in most of our notebooks, as they are currently super slow when testing the package.

henrymoss replied:
In the BoTorch notebooks they use 
`SMOKE_TEST = os.environ.get("SMOKE_TEST")`
and then write things like
`candidate_set_size = 1000 if not SMOKE_TEST else 5`

henrymoss replied:
they actually tell us how to do it https://github.com/pytorch/botorch/commit/5629ac4201cfdb1e3f51d42e25d190150f8c8dd9

apaleyes replied:
Was pondering the same question waiting for the builds to run. Will research

apaleyes replied:
So, we can use env variable, or even [papermill](https://github.com/yaananth/papermill) which allows to parametrize notebooks. But we will have to add something ugly in the tutorial, along these lines:
```
if "async-bo-enable-sleep" in os.environ:
    # long comment explaining the reader why this is here, about integ tests and what not
    enable_sleep_delays = os.getenv("async-bo-enable-sleep")
else:
    enable_sleep_delays = True
```

i find this ugly in a tutorial. Shall we just downscale the delay?

henrymoss replied:
we could just have one single environmental variable that can be used to scale down the runtimes for all notebooks (not a specific "async-bo-enable-sleep")

henrymoss replied:
if we name the variable "running_CI" or something similar, then we dont need a comment to explain whats going on.

henrymoss replied:
I would vote for having this line at the start of all our notebooks, as they are taking 20 mins to run at the moment. Any thoughts @uri-granta @hstojic @vpicheny @sebastianober ???

apaleyes replied:
i'll create an issue to discuss, it is definitely a good question, one that deserves a separate discussion

apaleyes replied:
#371

henrymoss reviewed (commented):
Getting there! Notebook is much better

vdutor commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+# %%
+def objective(points, sleep=True):
+    if points.shape[1] != 2:
+        raise ValueError(f"Incorrect input shape, expected (*, 2), got {points.shape}")
```
Minor: following TensorFlow syntax for tensor shapes.
```suggestion
        raise ValueError(f"Incorrect input shape, expected [None, 2], got {points.shape}")
```

vdutor commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+objective(np.array([[0.1, 0.5]]), sleep=False)
+
+# %% [markdown]
+# As always, we need to prepare model and some initial data to kick-start the optimization process.
```
```suggestion
# As always, we need to prepare the model and some initial data to kick-start the optimization process.
```

vdutor commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+#
+# To keep this notebook as reproducible as possible, we will only be using Python's multiprocessing package here. In this section we will explain our setup and define some common code to be used later.
+#
+# In both synchronous and asynchronous scenarios we will have a fixed set of worker processes performing observations. We will also have a main process responsible for optimization process with Trieste. When Trieste suggests a new point, it is inserted into a points queue. One of the workers picks this point from the queue, performs the observation, and inserts the output into observations queue. ...
```
```suggestion
# In both synchronous and asynchronous scenarios we will have a fixed set of worker processes performing observations. We will also have a main process responsible for optimization process with Trieste. When Trieste suggests a new point, it is inserted into a points queue. One of the workers picks this point from the queue, performs the observation, and inserts the output into the observations queue. The main process then picks up the observation from the queue, at which moment it either waits for the rest of the points in the batch to come back (synchronous scenario) or immediately suggests a new point (asynchronous scenario). This process continues either for a certain number of iterations or until we accumulate necessary number of observations.
```

apaleyes replied:
"asynchronous" is tricky to spell out, thanks!

vdutor commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+# In both synchronous and asynchronous scenarios we will have a fixed set of worker processes performing observations. We will also have a main process responsible for optimization process with Trieste. When Trieste suggests a new point, it is inserted into a points queue. One of the workers picks this point from the queue, performs the observation, and inserts the output into observations queue. ...
+# 
+# The overall setup is illustrated in this diagram:
+# ![multiprocessing setup](figures/async_bo.png)
```
Figure is not displayed in ipynb.

apaleyes replied:
Weird, it does for me. I'll build the docs locally to verify

apaleyes replied:
After local build image and all links render properly.

vdutor commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+
+# %% [markdown]
+# ## Asynchronous optimization
+# This section runs the asynchronous optimization routine. We first setup the [ask/tell optimizer](ask_tell_optimization.ipynb). Notice that we use **AsyncEfficientGlobalOptimization** rule specifically designed for asynchronous scenarios. Next we create thread-safe queues for points and observations, and run the optimization loop.
```
The "ask/tell optimizer" link doesn't seem to point anything?

apaleyes replied:
Unfortunate side effect of using pct.py instead of ipynb files in the codebase. The link will work in the final render of the docs! There is a tutorial on using ask/tell in Trieste which we refer to here.

vdutor commented on docs/notebooks/asynchronous_bayesian_optimization.pct.py:
```diff
@@ -0,0 +1,373 @@
+
+# %% [markdown]
+# ## Asynchronous optimization
+# This section runs the asynchronous optimization routine. We first setup the [ask/tell optimizer](ask_tell_optimization.ipynb). Notice that we use **AsyncEfficientGlobalOptimization** rule specifically designed for asynchronous scenarios. Next we create thread-safe queues for points and observations, and run the optimization loop.
```
Suggestion: a sentence on what the AsyncEfficientGlobalOptimization does differently could be interesting.

apaleyes replied:
good point! this is now re-written slightly because we renamed it to make it obvious it only works with greedy functions. hopefully the re-write made it clearer

vdutor reviewed (commented):
*Only went through the notebook.* Looks really good to me - only left a few minor suggestions.

apaleyes committed:
Address review comments

apaleyes committed:
Format

apaleyes commented:
Hopefully all the comments are addressed. Noticeable updates:
* Renamed rule to AsynchronousGreedy. No abbreviations, no shortcuts, only hardcore!
* Removed batch size parameter.
* I've run into [this bug](https://stackoverflow.com/questions/64462924/how-do-i-fix-raise-valueerrorduplicate-signature-r-signature-when-i-try) when building docs locally, and had to update version of jupyter_client (and resolve a conflict)
* Many-many fixes and updates in notebook's write-up

apaleyes committed:
Added note on potential performance improvement

hstojic reviewed (approved):
notebook looks great!

rule.py and function.py hopefully others were able to properly cover

apaleyes committed:
Merge branch 'develop' into async-bo
